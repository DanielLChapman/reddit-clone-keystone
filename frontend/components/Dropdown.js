import React, { useState } from 'react';
import PropTypes from 'prop-types'

function Dropdown(props) {
    /*
        let options = ['Hot', 'Best', 'New', 'Top', 'Rising'];
        const [dropValue, setDropValue] = useState('Hot');
        <Dropdown returnFunc={setDropValue} selected={dropValue} options={options} />    
    */
    return (
        <select value={props.selected} onChange={(e) => {
            props.returnFunc(e.target.value);
        }}>
            {
                props.options && props.options.map((x, i) => (
                    <option key={i} value={x}>{x}</option>
                ))
            }
        </select>
    );
}

Dropdown.propTypes = {
    returnFunc: PropTypes.func.isRequired,
    options: PropTypes.array.isRequired,
    selected: PropTypes.string.isRequired,
};
  

export default Dropdown;