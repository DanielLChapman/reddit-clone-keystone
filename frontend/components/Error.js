import React from 'react';

function Error(props) {
    return (
        <span>
            {props.error}
        </span>
    );
}

export default Error;