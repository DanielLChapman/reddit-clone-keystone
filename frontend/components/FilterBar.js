import React, { Component } from 'react';
import { FaHotjar, FaBahai, FaGopuram, FaChartLine} from 'react-icons/fa';



class FilterBar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            li1: true,
            li2: false,
            li3: false,
            li4: false
        }
    }

    toggleClass = (num) => {
        let state = {
            li1: false,
            li2: false,
            li3: false,
            li4: false
        };
        state['li'+num] = true;
        this.setState({...state});
    }

    render() {
        let style = {
            display: "none"
        };
        if (this.state.li3) {
            style= {
                display: "inline-block"
            };
        }
        return (
            <section className="filter-bar"> 
                <h5>Popular Posts</h5>
                <ul>
                    <li className={this.state.li1 ? 'active' : null} onClick={() => this.toggleClass(1)}><FaHotjar /> Hot</li>
                    <li className={this.state.li2 ? 'active' : null} onClick={() => this.toggleClass(2)}><FaBahai /> New</li>
                    <li className={this.state.li3 ? 'active' : null} onClick={() => this.toggleClass(3)}><FaGopuram /> Top</li>
                    <li className="time-selector-for-filter-bar" style={style}>Today</li>
                    <li className={this.state.li4 ? 'active' : null} onClick={() => this.toggleClass(4)}><FaChartLine /> Rising</li>
                </ul>
            </section>
        );
    }
}

export default FilterBar;