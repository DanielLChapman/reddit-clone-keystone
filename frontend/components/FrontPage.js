import Link from 'next/link';
import React from 'react';
import { useUser } from './User';

function FrontPage(props) {
    const user = useUser();
    return (
        <div>
            {user && (
                <Link href="/subreddits/create">Create Your Own Subreddit</Link>
            )}
            
        </div>
    );
}

export default FrontPage;