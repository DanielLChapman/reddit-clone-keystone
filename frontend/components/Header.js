import Link from 'next/link';
import Nav from './Nav';


export default function Header() {
  return (
    <>
      <div className="bar">
        <span>
          <Link href="/">App name</Link>
        </span>
        <Nav />
      </div>
    </>
  );
}
