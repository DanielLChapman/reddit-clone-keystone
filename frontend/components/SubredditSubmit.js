import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import React from 'react';
import useForm from '../lib/useForm';
import Router from 'next/router';
import Form from './styles/SubredditForm';


//NEED TO SET OWNER HERE
export const CREATE_SUBREDDIT_MUTATION = gql`
    mutation CREATE_SUBREDDIT_MUTATION(
        $name: String!,
        $slug: String!,
        $description: String,
        $sidebar: String,
        $title: String!
    ) {
        createSubreddit(
            data: {
                name: $name,
                slug: $slug,
                description: $description,
                sidebar: $sidebar,
                title: $title
            }
        ) {
            id
            name
        }
    }
`;


function SubredditSubmitForm(props) {
    const { inputs, handleChange, resetForm } = useForm({
        name: '',
        slug: '',
        description: '',
        sidebar: '',
        title: '',
        errors: {
            name: 'false'
        }
      });

      const [createSubreddit, {loading, data, error}] = useMutation(CREATE_SUBREDDIT_MUTATION);


    
    return (
        <div>
            <Form
            onSubmit={async (e) => {
                e.preventDefault();
                //form validation
                let format = /[ `!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~]/;
                if (format.test(inputs.name)) {
                    alert('No spaces or special characters in name field');
                    let target = {
                        target: {
                            value:'true', 
                            name: 'name', 
                            type: 'error'
                        }
                    }
                    handleChange(target);
                    return;
                } 

                const res = await createSubreddit({
                    variables: {
                        name: inputs.name,
                        description: inputs.description,
                        sidebar: inputs.sidebar,
                        title: inputs.title,
                        slug: inputs.name.toLowerCase()
                    }
                })

                Router.push({
                    pathname: `/r/${inputs.name.toLowerCase()}`,
                  });

            }}>
                <fieldset>
                    <label htmlFor="name">
                        Name
                        <br />
                        <span className="lower-form-text">no spaces or special characters.</span>
                        <input 
                            required
                            type="text"
                            id="name"
                            name="name"
                            className={inputs.errors.name === 'true' ? 'error-field' : ''}
                            value={inputs.name}
                            placeholder="name"
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="title">
                        Title <br />
                        <input 
                            required
                            type="text"
                            id="title"
                            name="title"
                            placeholder="title"
                            value={inputs.title}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="description">
                        Description <br />
                        <textarea
                            id="description"
                            name="description"
                            placeholder="description"
                            value={inputs.description}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="sidebar">
                        Sidebar <br />
                        <textarea
                            id="sidebar"
                            name="sidebar"
                            placeholder="sidebar"
                            value={inputs.sidebar}
                            onChange={handleChange}
                        />
                    </label>
                    <button type="submit">Create Subreddit</button>
                    
                </fieldset>

            </Form>
        </div>
    );
}

export default SubredditSubmitForm;