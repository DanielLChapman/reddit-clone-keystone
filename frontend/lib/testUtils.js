import casual from 'casual';

  
// seed it so we get consistent results
casual.seed(777);

const fakeUser = (overrides) => (
    {
    __typename: 'User',
    id: '4234',
    name: casual.name,
    username: casual.name,
    owner: [],
    email: casual.email,
    permissions: ['ADMIN'],
    moderating: [],
    subreddits: [],
    posts: [],
    postvotes: [],
    firstSignin: false,
    comments: [],
    commentvotes:[],
    ...overrides,
});

const fakeSubreddit = (overrides) => (
    {
        __typename: 'Subreddit',
        id: casual.uuid,
        name: casual.string,
        title: casual.string,
        slug: casual.string,
        sidebar: casual.string,
        rules: casual.string,
        description: casual.string,
        status: 'PUBLIC',
        ownder: fakeUser(),
        moderators: [],
        subscriber: [],
        posts: [],
        createdAt: Date.now(),
    }
);

const fakeLinkPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'www.google.com',
    type: 'Link',
    comments: [],
    removed: 'False',
    id: casual.uuid,


});

const fakeMediaPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'https://cdn.fileinfo.com/img/ss/sm/png_79.jpg',
    type: 'Link',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeTextPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    content: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: '',
    type: 'Text',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeComment = (overrides) => (
    {
    __typename: 'Comment',
    content: casual.sentence,
    user: fakeUser(),
    post: fakeTextPost(),
    createdAt: Date.now(),
    votes: [],
    parent: null,
    id: casual.uuid,
    children: [],

});


// Fake LocalStorage
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  }



export {

    LocalStorageMock,
    fakeLinkPost,
    fakeComment,
    fakeMediaPost,
    fakeSubreddit,
    fakeTextPost,
    fakeUser,

  };