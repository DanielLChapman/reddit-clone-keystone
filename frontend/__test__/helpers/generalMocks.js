import { CURRENT_USER_QUERY } from '../../components/User';
import { fakeTextPost, fakeUser} from '../../lib/testUtils';

const fakeUserObj = fakeUser();
const notSignedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: null } },
    },
  ];
  
const signedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: fakeUser() /*{
        
            id: fakeUserObj.id,
            username: fakeUserObj.username,
            email: fakeUserObj.email,
            name:fakeUserObj.name,
            posts: [],
            comments: [],
            moderating: [],
            owner:[],
            subreddits:[],
            postvotes:[],
            commentvotes: [],
        
      }*/ } },
    },
  ];
  
const signedInMocksWithPosts = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: {
        data: {
          authenticatedItem: fakeUser({
            posts: [fakeTextPost()],
          }),
        },
      },
    },
];

export {
    notSignedInMocks,
    signedInMocks,
    signedInMocksWithPosts
}

/* 

{
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }

    */