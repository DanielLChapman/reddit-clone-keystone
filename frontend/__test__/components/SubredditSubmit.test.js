import { act, render, screen, waitFor } from '@testing-library/react';
import {MockedProvider} from '@apollo/react-testing';
import { signedInMocks, signedInMocksWithPosts } from '../helpers/generalMocks';
import wait from 'waait';
import SubredditCreatePage from '../../pages/subreddits/create';
import { fakeSubreddit } from '../../lib/testUtils';
import userEvent from '@testing-library/user-event';
import { CREATE_SUBREDDIT_MUTATION } from '../../components/SubredditSubmit';
import  Router  from 'next/router';

const subreddit = fakeSubreddit();

jest.mock('next/router', () => ({
    push: jest.fn(),
  }));

const mocks = [
    // Mutation Mock
    ...signedInMocks, 
    {
      request: {
        query: CREATE_SUBREDDIT_MUTATION,
        variables: {
          name: subreddit.name,
          description: subreddit.description,
          sidebar: subreddit.sidebar,
          title: subreddit.title,
          slug: subreddit.name.toLowerCase(),
        },
      },
      result: {
        data: {
          createSubreddit: {
            id: subreddit.id,
            __typename: 'Subreddit',
            name: subreddit.name,
          },
        },
      },
    },

  ];

describe('it renders a subreddit submit page and submits successfully', () => {
    it ('does not load unless signed in', async () => {
        const { container, debug } = render(
            <MockedProvider >
              <SubredditCreatePage />
            </MockedProvider>
        );

        await screen.findByText(
            'Sign Into Your Account'
        );

        expect(container).toMatchSnapshot();
    });

    it ('loads when signed in', async () => {
        const { container, debug } = render(
            <MockedProvider mocks={signedInMocks}>
              <SubredditCreatePage />
            </MockedProvider>
        );

        await screen.findByText(
            'Create Subreddit'
        );

        expect(container).toMatchSnapshot();
    });

    it ('calls the mutation', async () => {
        const { container, debug } = render(
            <MockedProvider mocks={mocks}>
              <SubredditCreatePage />
            </MockedProvider>
        );

        await screen.findByText(
            'Create Subreddit'
        );

        await userEvent.type(screen.getByPlaceholderText('name'), subreddit.name);
        await userEvent.type(screen.getByPlaceholderText('title'), subreddit.title);
        await userEvent.type(screen.getByPlaceholderText('description'), subreddit.description);
        await userEvent.type(screen.getByPlaceholderText('sidebar'), subreddit.sidebar);

        await userEvent.click(screen.getByText('Create Subreddit'));

        await waitFor(() => wait(0));
        expect(Router.push).toHaveBeenCalled();
    });
})