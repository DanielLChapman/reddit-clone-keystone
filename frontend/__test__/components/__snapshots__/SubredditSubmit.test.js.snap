// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a subreddit submit page and submits successfully does not load unless signed in 1`] = `
<div>
  <div>
    <form
      class="Form-sc-1tx35ym-0 etUUQp"
      method="POST"
    >
      <h2>
        Sign Into Your Account
      </h2>
      <fieldset>
        <label
          for="email"
        >
          Email
          <input
            autocomplete="email"
            name="email"
            placeholder="Your Email Address"
            type="email"
            value=""
          />
        </label>
        <label
          for="password"
        >
          Password
          <input
            autocomplete="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </label>
        <button
          type="submit"
        >
          Sign In
        </button>
      </fieldset>
    </form>
    <style>
                      body, html{                background-color:white;                }            
    </style>
  </div>
</div>
`;

exports[`it renders a subreddit submit page and submits successfully loads when signed in 1`] = `
<div>
  <div>
    <div>
      <form
        class="SubredditForm__Form-q7oa8i-0 kqHjhW"
      >
        <fieldset>
          <label
            for="name"
          >
            Name 
            false
            <br />
            <span
              class="lower-form-text"
            >
              no spaces or special characters.
            </span>
            <input
              class=""
              id="name"
              name="name"
              placeholder="name"
              required=""
              type="text"
              value=""
            />
          </label>
          <label
            for="title"
          >
            Title 
            <br />
            <input
              id="title"
              name="title"
              placeholder="title"
              required=""
              type="text"
              value=""
            />
          </label>
          <label
            for="description"
          >
            Description 
            <br />
            <textarea
              id="description"
              name="description"
              placeholder="description"
            />
          </label>
          <label
            for="sidebar"
          >
            Sidebar 
            <br />
            <textarea
              id="sidebar"
              name="sidebar"
              placeholder="sidebar"
            />
          </label>
          <button
            type="submit"
          >
            Create Subreddit
          </button>
        </fieldset>
      </form>
    </div>
    <style>
                      body, html{                background-color:white;                }            
    </style>
  </div>
</div>
`;
